### ISSUE TITLE
Bug: Quartz going in infinite loop 

### ISSUE THREAD

**Issue by dorinsimionescu** [quartznet/quartznet#612]
### Issue
Due to DST, fire times for Daily trigger are not correctly calculated, eg they are calculated and fired in the past. This causes misfire and further more, the trigger will never advance to the next date (it loops between 24 and 25 of March).

### Expected behavior
Trigger should fire according to set schedule, eg next fire date should be on the next day, at 00:00.
### Actual behavior
The scheduler goes into an infinite loop and all dates are set in the past. Trigger continuously misfires.

### Steps to reproduce
Change system date to 25/03/2018 (from cmd, date 25/03/2018)
Change system time to 22:10:10 (from cmd, time 22:10:00)
Set timezone to UTC + 2 Athens Bucharest
Create a daily schedule that fires between 00:00 and 22:00

The bug does not reproduce by usingSystemTime.Utc and SystemTime.UtcNow functions, you need to actually change the PC date and time. I tested on Windows 10 Enterprise, but it has the same behaviour on Windows Server. 

### Code to reproduce the issue after you change system date, time and timezone:

```csharp

namespace TestQuartz
{
    public class HelloJob : IJob
    {
        public async Task Execute(IJobExecutionContext context)
        {
            await Console.Out.WriteLineAsync( string.Format("Greetings from HelloJob! FireTimeUtc: {0} NextFireTime: {1}, StartTimeUtc: {2}", 
                context.FireTimeUtc, 
                context.Trigger.GetNextFireTimeUtc(),
                context.Trigger.StartTimeUtc
                
                ));

        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            RunProgram().GetAwaiter().GetResult();
        }

        private static async Task RunProgram()
        {
            try
            {

                // Grab the Scheduler instance from the Factory
                NameValueCollection props = new NameValueCollection
                {
                    { "quartz.serializer.type", "binary" }
                };
                StdSchedulerFactory factory = new StdSchedulerFactory(props);
                IScheduler scheduler = await factory.GetScheduler();


                // and start it off
                await scheduler.Start();

                // some sleep to show what's happening
                // define the job and tie it to our HelloJob class
                IJobDetail job = JobBuilder.Create<HelloJob>()
                    .WithIdentity("job1", "group1")
                    .Build();

                // Trigger the job to run now, and then repeat every 10 seconds
                var builder = TriggerBuilder.Create()
                    .WithIdentity("trigger1", "group1")
                ;

                var trigger =
                    builder
                     .WithDailyTimeIntervalSchedule(x => x
                    .InTimeZone(TimeZoneInfo.Local)
                    .StartingDailyAt(new TimeOfDay(0, 0, 0))
                    .EndingDailyAt(new TimeOfDay(22, 0, 0))
                    .WithInterval(15, IntervalUnit.Minute)
                    .WithMisfireHandlingInstructionDoNothing()
                    )
                    .WithPriority(5)
                    .Build();


                var times2 = TriggerUtils.ComputeFireTimes((Quartz.Spi.IOperableTrigger)trigger, null, 2000);
                // check the times..


                // Tell quartz to schedule the job using our trigger
                await scheduler.ScheduleJob(job, trigger);

                // some sleep to show what's happening
                await Task.Delay(TimeSpan.FromSeconds(60));

                // and last shut down the scheduler when you are ready to close your program
                await scheduler.Shutdown();
            }
            catch (SchedulerException se)
            {
                await Console.Error.WriteLineAsync(se.ToString());
            }
        }
    }
}

```

### Proposed fix
As far as we saw, the issue is in AdvanceToNextDayOfWeekIfNecessary from the class DailyTimeIntervalTriggerImpl. 
Maybe a fix is : 
```csharp
if (daysOfWeek.Contains(dayOfWeekOfFireTime))
{
    fireTime = CreateCalendarTime(fireTimeStartDate); // instead of fireTimeStartDateCal;
    break;
}
```

**Comment by dougajmcdonald** [comment]
We saw this same issue with the recent DST change and would be very interested in a solution!

**Comment by lahma** [comment]
Let's try to get this fixed soon. I'm currently on winter vacation (which usually means I can spare time), but this has to wait until I'm back unless somebody would like to do a PR backed with tests. That would speed things up.

**Comment by dradoaica** [comment]
I am facing the same issue!

**Comment by dradoaica** [comment]
I compared the "DailyTimeIntervalTriggerImpl.cs" file with the "DailyTimeIntervalTriggerImpl.java" file from the Java implementation and it seems you missed something:

```csharp
    public override DateTimeOffset? GetFireTimeAfter(DateTimeOffset? afterTime)
    {
        ...
        // f. Continue to calculate the fireTime by incremental unit of intervals.
        // recall that if fireTime was less that fireTimeStartDate, we didn't get this far
        startTimeUtc = TimeZoneUtil.ConvertTime(fireTimeStartDate, TimeZone);
        long secondsAfterStart = (long) (fireTime.Value - startTimeUtc).TotalSeconds;
        long repeatLong = RepeatInterval;

        DateTimeOffset sTime = fireTimeStartDate.ToUniversalTime(); // here it shold be CreateCalendarTime(...)!!!
        IntervalUnit repeatUnit = RepeatIntervalUnit;
        if (repeatUnit == IntervalUnit.Second)
        {
            long jumpCount = secondsAfterStart / repeatLong;
        ...
    }
```
according to:

```java
    @Override
    public Date getFireTimeAfter(Date afterTime) {
        ...
        // f. Continue to calculate the fireTime by incremental unit of intervals.
        // recall that if fireTime was less that fireTimeStartDate, we didn't get this far
        long fireMillis = fireTime.getTime();
        long startMillis = fireTimeStartDate.getTime();
        long secondsAfterStart = (fireMillis - startMillis) / 1000L;
        long repeatLong = getRepeatInterval();
        Calendar sTime = createCalendarTime(fireTimeStartDate);
        IntervalUnit repeatUnit = getRepeatIntervalUnit();
        if(repeatUnit.equals(IntervalUnit.SECOND)) {
            long jumpCount = secondsAfterStart / repeatLong;
        ...
    }
```

**Comment by lahma** [comment]
Fix will be part of 3.0.5 thanks to @dradoaica 's PR.
